File: java/src/main/java/io/hron/ast/Weekday.java
package io.hron.ast;

import java.util.Map;
import java.util.Optional;

/** Represents a day of the week. */
public enum Weekday {
  MONDAY(1, "monday"),
  TUESDAY(2, "tuesday"),
  WEDNESDAY(3, "wednesday"),
  THURSDAY(4, "thursday"),
  FRIDAY(5, "friday"),
  SATURDAY(6, "saturday"),
  SUNDAY(7, "sunday");

  private final int isoNumber;
  private final String displayName;

  Weekday(int isoNumber, String displayName) {
    this.isoNumber = isoNumber;
    this.displayName = displayName;
  }

  /**
   * Returns the ISO 8601 day number (Monday=1, Sunday=7).
   *
   * @return the ISO day number
   */
  public int number() {
    return isoNumber;
  }

  /**
   * Returns the cron day of week number (Sunday=0, Monday=1, ..., Saturday=6).
   *
   * @return the cron day of week number
   */
  public int cronDOW() {
    return switch (this) {
      case SUNDAY -> 0;
      case MONDAY -> 1;
      case TUESDAY -> 2;
      case WEDNESDAY -> 3;
      case THURSDAY -> 4;
      case FRIDAY -> 5;
      case SATURDAY -> 6;
    };
  }

  @Override
  public String toString() {
    return displayName;
  }

  private static final Map<String, Weekday> PARSE_MAP =
      Map.ofEntries(
          Map.entry("monday", MONDAY), Map.entry("mon", MONDAY),
          Map.entry("tuesday", TUESDAY), Map.entry("tue", TUESDAY),
          Map.entry("wednesday", WEDNESDAY), Map.entry("wed", WEDNESDAY),
          Map.entry("thursday", THURSDAY), Map.entry("thu", THURSDAY),
          Map.entry("friday", FRIDAY), Map.entry("fri", FRIDAY),
          Map.entry("saturday", SATURDAY), Map.entry("sat", SATURDAY),
          Map.entry("sunday", SUNDAY), Map.entry("sun", SUNDAY));

  /**
   * Parses a weekday name (case insensitive).
   *
   * @param s the string to parse
   * @return the weekday if valid
   */
  public static Optional<Weekday> parse(String s) {
    return Optional.ofNullable(PARSE_MAP.get(s.toLowerCase()));
  }

  /**
   * Returns a Weekday from an ISO 8601 day number.
   *
   * @param n the ISO day number (1-7)
   * @return the weekday if valid
   */
  public static Optional<Weekday> fromNumber(int n) {
    if (n < 1 || n > 7) {
      return Optional.empty();
    }
    return Optional.of(values()[n - 1]);
  }

  /**
   * Returns a Weekday from a java.time.DayOfWeek.
   *
   * @param dow the DayOfWeek
   * @return the corresponding Weekday
   */
  public static Weekday fromDayOfWeek(java.time.DayOfWeek dow) {
    return values()[dow.getValue() - 1];
  }

  /**
   * Converts this Weekday to a java.time.DayOfWeek.
   *
   * @return the corresponding DayOfWeek
   */
  public java.time.DayOfWeek toDayOfWeek() {
    return java.time.DayOfWeek.of(isoNumber);
  }
}

File: csharp/Hron/Ast/Weekday.cs
namespace Hron.Ast;

/// <summary>
/// Represents a day of the week.
/// </summary>
public enum Weekday
{
    Monday = 1,
    Tuesday = 2,
    Wednesday = 3,
    Thursday = 4,
    Friday = 5,
    Saturday = 6,
    Sunday = 7
}

public static class WeekdayExtensions
{
    private static readonly Dictionary<string, Weekday> ParseMap = new(StringComparer.OrdinalIgnoreCase)
    {
        ["monday"] = Weekday.Monday,
        ["mon"] = Weekday.Monday,
        ["tuesday"] = Weekday.Tuesday,
        ["tue"] = Weekday.Tuesday,
        ["wednesday"] = Weekday.Wednesday,
        ["wed"] = Weekday.Wednesday,
        ["thursday"] = Weekday.Thursday,
        ["thu"] = Weekday.Thursday,
        ["friday"] = Weekday.Friday,
        ["fri"] = Weekday.Friday,
        ["saturday"] = Weekday.Saturday,
        ["sat"] = Weekday.Saturday,
        ["sunday"] = Weekday.Sunday,
        ["sun"] = Weekday.Sunday
    };

    /// <summary>
    /// Returns the ISO 8601 day number (Monday=1, Sunday=7).
    /// </summary>
    public static int Number(this Weekday weekday) => (int)weekday;

    /// <summary>
    /// Returns the cron day of week number (Sunday=0, Monday=1, ..., Saturday=6).
    /// </summary>
    public static int CronDOW(this Weekday weekday) => weekday switch
    {
        Weekday.Sunday => 0,
        Weekday.Monday => 1,
        Weekday.Tuesday => 2,
        Weekday.Wednesday => 3,
        Weekday.Thursday => 4,
        Weekday.Friday => 5,
        Weekday.Saturday => 6,
        _ => throw new ArgumentOutOfRangeException(nameof(weekday))
    };

    /// <summary>
    /// Returns the lowercase display name.
    /// </summary>
    public static string ToDisplayString(this Weekday weekday) => weekday.ToString().ToLowerInvariant();

    /// <summary>
    /// Parses a weekday name (case insensitive).
    /// </summary>
    public static Weekday? Parse(string s)
        => ParseMap.TryGetValue(s, out var weekday) ? weekday : null;

    /// <summary>
    /// Returns a Weekday from an ISO 8601 day number.
    /// </summary>
    public static Weekday? FromNumber(int n)
        => n is >= 1 and <= 7 ? (Weekday)n : null;

    /// <summary>
    /// Returns a Weekday from a DayOfWeek.
    /// </summary>
    public static Weekday FromDayOfWeek(DayOfWeek dow) => dow switch
    {
        DayOfWeek.Monday => Weekday.Monday,
        DayOfWeek.Tuesday => Weekday.Tuesday,
        DayOfWeek.Wednesday => Weekday.Wednesday,
        DayOfWeek.Thursday => Weekday.Thursday,
        DayOfWeek.Friday => Weekday.Friday,
        DayOfWeek.Saturday => Weekday.Saturday,
        DayOfWeek.Sunday => Weekday.Sunday,
        _ => throw new ArgumentOutOfRangeException(nameof(dow))
    };

    /// <summary>
    /// Converts this Weekday to a DayOfWeek.
    /// </summary>
    public static DayOfWeek ToDayOfWeek(this Weekday weekday) => weekday switch
    {
        Weekday.Monday => DayOfWeek.Monday,
        Weekday.Tuesday => DayOfWeek.Tuesday,
        Weekday.Wednesday => DayOfWeek.Wednesday,
        Weekday.Thursday => DayOfWeek.Thursday,
        Weekday.Friday => DayOfWeek.Friday,
        Weekday.Saturday => DayOfWeek.Saturday,
        Weekday.Sunday => DayOfWeek.Sunday,
        _ => throw new ArgumentOutOfRangeException(nameof(weekday))
    };
}
